### PR TITLE
Added optional setting http_proxy to push_image

### PR DESCRIPTION
**Pull request by xNUTx**
This will allow the user to set up an **external reverse proxy** to have the HA instance operate on HTTPS and push_image use that reverse proxy to avoid using HTTPS traffic processing on the plate.

Besides now working around a bug in the libraries that kills SSL traffic on the plates, it will also allow the plate not to deal with SSL encrypted connections at all (even when it is working again) to save processing power.

### CODE REVIEW

**Comment by dgomes** [comment]
Naming this "http_proxy" is misleading, what you are doing is defining the HA as another hostname

**Comment by fvanroie** [comment]
> Naming this "http_proxy" is misleading, what you are doing is defining the HA as another hostname

So, maybe `ALIAS` is a better moniker?

**Comment by dgomes** [comment]
Why don't you move the SSL to the reverse proxy ? and leave the internal HA without SSL ? (that's actual the most common approach)

**Comment by xNUTx** [comment]
It is not misleading, no more as the proxy settings in a Linux shell or in a web browser. It is a designated address you divert the request to to proxy it to a different protocol.

I have my HA instance reachable from the web and everywhere in the app, special port and https only, to make sure everything works I use this address locally and externally. The only way to do it secure without using the cloud services from nabu casa.

**Comment by dgomes** [comment]
proxy settings in a Linux shell or browser are used by the client, your OpenHASP plate is the client... not HA (this component is the server) - that is why naming it http_proxy is misleading.

If you run a local reverse proxy, the proper way is to move SSL termination to the reverse proxy and offload SSL from HA. This way your internal devices can access HA without SSL while external access uses SSL https://www.home-assistant.io/integrations/http/#reverse-proxies

**Comment by xNUTx** [comment]
> proxy settings in a Linux shell or browser are used by the client, your OpenHASP plate is the client... not HA (this component is the server) - that is why naming it http_proxy is misleading.

This is not entirely accurate, from my viewpoint: The integration is sending the plate an image configuration (width, height, url and a command to start loading that image). Which then gets used by the plate as cliënt to execute the desired action. As such, what you do with the HA integration in the automation is configuring the client to use a proxy. Just like you configure it to load specified image.

The reverse proxy in this case is on the local network, nobody and nothing will be able to use it from the outside. I think the solution I have set up is generally more secure then to use a proxy to offload ssl traffic for the outside world. The way I did it, every connection to HA is secured. The proxy is configured to allow only access to the plates. Nothing else can use it.

**Comment by fvanroie** [comment]
Is this still active?

**Comment by xNUTx** [comment]
I would definitely prefer to have this in a next release, yes.

It's just an alternative access for https hosted images... 

Op 12 mei 2024 14:54, om 14:54, fvanroie ***@***.***> schreef:

**Comment by xNUTx** [comment]
It required no modifications for the merge with the release version.

File: custom_components/openhasp/__init__.py
Comment (xNUTx) [reply custom_components/openhasp/__init__.py]: Yah, you're not wrong. This has been fixed :)

File: custom_components/openhasp/__init__.py
Comment (xNUTx) [reply custom_components/openhasp/__init__.py]: PHP in the way of the Python syntax. Sorry :)

File: custom_components/openhasp/__init__.py
Comment (dgomes) [inline custom_components/openhasp/__init__.py:559]: This is unrelated to this PR, if there is an issue please open a separate PR

File: custom_components/openhasp/__init__.py
Comment (dgomes) [inline custom_components/openhasp/__init__.py:606]: instead of testing against None change the if order

File: custom_components/openhasp/services.yaml
Comment (dgomes) [inline custom_components/openhasp/services.yaml:111]:
```suggestion
      description: Proxy address to use. This can be used to allow HTTP access to an otherwise SSL secured HA instance. Offering the proxy functionality is out of the scope of this integration.
```

**Review by dgomes — CHANGES_REQUESTED** [review]
I think http_proxy is not the write term... I would call it external URL

File: custom_components/openhasp/__init__.py
Comment (xNUTx) [reply custom_components/openhasp/__init__.py:559]: I intended to, but the change got included anyway... unfortunate way of working in the same branch.. you can cherry pick it out and I will load it into another once this one is merged.

There is an issue that is fixed with this change, the display stops or refuses to dynamically update and throws an error in the logs because (most likely) num_pages contains a string, while page contains an int, resulting in a python type error.

Comment (dgomes) [reply custom_components/openhasp/__init__.py:559]: I can't cherry pick on the PR... please make a commit to revert this change

**Comment by xNUTx** [comment]
> I think http_proxy is not the write term... I would call it external URL

The thing is though... it is not an external URL at all. It is meant to be used within the secured environment.

With HA running with a https connection secured with a certificate, it will no longer listen on HTTP. That is why the **protocol proxy** is needed to allow HTTP access to an otherwise HTTPS secured location of an image. This is required to allow the plate to load a dynamic image source from a different HA component (in my case, it loads the image of the doorbell camera).

We went over this... I thought I was clear enough in my explanation before.

File: custom_components/openhasp/__init__.py
Comment (xNUTx) [reply custom_components/openhasp/__init__.py:606]: Python is not my strongsuit, I understand your request but I am not sure (but suspect) if I can remove the '== None' and rely on python to understand that anything but 'None' is going to be okay to use... (with the actions reversed, obviously)

Also, it is a syntax thing, while it is perfectly functional... but if you can confirm my suspicion above, I will fix it.

**Comment by dgomes** [comment]
> > I think http_proxy is not the write term... I would call it external URL
> 
> The thing is though... it is not an external URL at all. It is meant to be used within the secured environment.
> 
> With HA running with a https connection secured with a certificate, it will no longer listen on HTTP. That is why the **protocol proxy** is needed to allow HTTP access to an otherwise HTTPS secured location of an image. This is required to allow the plate to load a dynamic image source from a different HA component (in my case, it loads the image of the doorbell camera).
> 
> We went over this... I thought I was clear enough in my explanation before.

I never said I agree :) but I leave to @fvanroie to make the final call

File: custom_components/openhasp/__init__.py
Comment (xNUTx) [reply custom_components/openhasp/__init__.py]: ok, change applied :)

File: custom_components/openhasp/__init__.py
Comment (xNUTx) [reply custom_components/openhasp/__init__.py:559]: Ok, will do.

**Pull request closed without merging**
